hypervisor/arch/x86/page.c

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (C) 2018 Intel Corporation. All rights reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+#include <hypervisor.h>
+
+#define PML4_PAGE_NUM(size)	1UL
+#define PDPT_PAGE_NUM(size)	(((size) + PML4E_SIZE - 1UL) >> PML4E_SHIFT)
+#define PD_PAGE_NUM(size)	(((size) + PDPTE_SIZE - 1UL) >> PDPTE_SHIFT)
+#define PT_PAGE_NUM(size)	(((size) + PDE_SIZE - 1UL) >> PDE_SHIFT)
+
+#define DEFINE_PGTABLE_PAGE(prefix, lvl, LVL, size)	\
+		static struct page prefix ## lvl ## _pages[LVL ## _PAGE_NUM(size)]
+
+DEFINE_PGTABLE_PAGE(primary_, pml4, PML4, CONFIG_PLATFORM_RAM_SIZE + CONFIG_PLATFORM_MMIO_SIZE);
+DEFINE_PGTABLE_PAGE(primary_, pdpt, PDPT, CONFIG_PLATFORM_RAM_SIZE + CONFIG_PLATFORM_MMIO_SIZE);
+DEFINE_PGTABLE_PAGE(primary_, pd, PD, CONFIG_PLATFORM_RAM_SIZE + CONFIG_PLATFORM_MMIO_SIZE);
+
+/* ppt: pripary page table */
+static union pgtable_pages_info ppt_pages_info = {
+	.ppt = {
+		.pml4_base = primary_pml4_pages,
+		.pdpt_base = primary_pdpt_pages,
+		.pd_base = primary_pd_pages,
+	}
+};
+
+static inline uint64_t get_primary_default_access_right(void)
+{
+	return (PAGE_PRESENT | PAGE_RW | PAGE_USER);
+}
+
+static inline uint64_t primary_pgentry_present(uint64_t pte)
+{
+	return pte & PAGE_PRESENT;
+}
+
+static inline struct page *get_primary_pml4_page(const union pgtable_pages_info *info, __unused uint64_t gpa)
```

Comment:
Now we have both ppt & primary_xxx.   All use ppt?

Reply:
I think ppt means primary page table, while primary pml4 is one of it (page table could be pml4/pdpt/pd/pt). Can this make sense ?